File: module-web/src/main/java/maple/expectation/web/dto/page/CursorPageRequest.java
package maple.expectation.web.dto.page;

/** Cursor-based Pagination 요청 (#233) */
public record CursorPageRequest(Long cursor, int size) {
  private static final int DEFAULT_SIZE = 20;
  private static final int MAX_SIZE = 100;

  public static CursorPageRequest of(Long cursor, int size) {
    int validSize = size <= 0 ? DEFAULT_SIZE : Math.min(size, MAX_SIZE);
    return new CursorPageRequest(cursor, validSize);
  }

  public static CursorPageRequest firstPage() {
    return new CursorPageRequest(null, DEFAULT_SIZE);
  }

  public static CursorPageRequest firstPage(int size) {
    return of(null, size);
  }
}

File: module-web/src/main/kotlin/maple/expectation/web/dto/page/CursorPageRequest.kt
package maple.expectation.web.dto.page

/**
 * Cursor-based Pagination 요청 (#233)
 *
 * @param cursor 마지막으로 조회한 커서 (ID)
 * @param size 페이지 크기
 */
data class CursorPageRequest(
    val cursor: Long?,
    val size: Int
) {
    companion object {
        private const val DEFAULT_SIZE = 20
        private const val MAX_SIZE = 100

        @JvmStatic
        fun of(cursor: Long?, size: Int): CursorPageRequest {
            val validSize = if (size <= 0) DEFAULT_SIZE else minOf(size, MAX_SIZE)
            return CursorPageRequest(cursor, validSize)
        }

        @JvmStatic
        fun firstPage(): CursorPageRequest {
            return CursorPageRequest(null, DEFAULT_SIZE)
        }

        fun firstPage(size: Int): CursorPageRequest {
            return of(null, size)
        }
    }
}
